include/pmacc/math/complex/Bessel.tpp

```diff
@@ -0,0 +1,286 @@
+
+            float_T const a0 = pmMath::abs( z );
+            complex_T const z2 = z * z;
+            complex_T z1 = z;
+            if( a0 == float_T( 0.0 ) )
+            {
+                cj0 = cone;
+                return cj0;
+            }
+            if( z.get_real() < float_T( 0.0 ) )
+                z1 = float_T( -1.0 ) * z;
+            if( a0 <= float_T( 12.0 ) )
+            {
+                cj0 = cone;
+                complex_T cr = cone;
+                for ( uint32_t k = 1u; k <= 40u; k++ )
+                {
+                    cr *= float_T( -0.25 ) * z2 / float_T( k * k );
+                    cj0 += cr;
+                    if( pmMath::abs( cr ) < pmMath::abs( cj0 ) * eps ) break;
+                }
+            }
+            else {
+                uint32_t kz;
+                if( a0 >= float_T( 50.0 ) ) kz = 8u;       // can be changed to 10
+                else if( a0 >= float_T( 35.0 ) ) kz = 10u; //   "      "     "  12
+                else kz = 12u;                             //   "      "     "  14
+                complex_T ct1 = z1 - float_T( M_PI_4 );
+                complex_T cp0 = cone;
+                for ( uint32_t k = 0u; k < kz; k++ )
+                {
+                    cp0 += a[ k ] * pow(
+                        z1,
+                        float_T( -2.0 ) * k - float_T( 2.0 )
+                    );
+                }
+                complex_T cq0 = float_T( -0.125 ) / z1;
+                for ( uint32_t k = 0; k < kz; k++ )
+                {
+                    cq0 += b[ k ] * pmMath::pow(
+                        z1,
+                        float_T( -2.0 ) * k - float_T( 3.0 )
+                    );
+                }
+                complex_T const cu = pmMath::sqrt( float_T( M_2_PI ) / z1 );
+                cj0 = cu * ( cp0 * pmMath::cos( ct1 ) - cq0 * pmMath::sin( ct1 ) );
+            }
+            return cj0;
+        }
+    };
+
+    template<
+        typename T_Type,
+        typename T_TableA,
+        typename T_TableB,
+        typename T_TableA1,
+        typename T_TableB1
+    >
+    struct Cbesselj1Base
+    {
+        using Result = pmacc::math::Complex< T_Type >;
+        using complex_T = pmacc::math::Complex< T_Type >;
+        using float_T = T_Type;
+
+        HDINLINE Result operator( )( complex_T const & z )
+        {
+            T_TableA a;
+            T_TableB b;
+            T_TableA1 a1;
+            T_TableB1 b1;
+            Result cj1;
+            /* The target rel. accuracy goal eps is chosen according to the original implementation
+             * of C. Bond, where for double-precision the accuracy goal is 1.0e-15. Here the accuracy
+             * goal value is the same 4.5 * DBL_EPSILON = 1.0e-15 for double-precision, but is similarly
+             * defined for float-precision.
+             */
+            float_T const eps = float_T( 4.5 ) * std::numeric_limits< float_T >::epsilon( );
```

Comment:
I missed that we need to enable the nvcc compiler flag `--relaxed-constexpr` to allow the usage of `std::numeric_limits` within a device function. Have you tested this code on GPU or CPU?

Reply:
PR to add the needed flag is in the queue https://github.com/ComputationalRadiationPhysics/alpaka/pull/416

Reply:
`std::numeric_limits` is now available on device side #2328